### ISSUE TITLE
Define permissions for resolver functions

### ISSUE THREAD

**Issue by marktani** [prismagraphql/graphcool-framework#120]
<a href="https://github.com/kbrandwijk"><img src="https://avatars1.githubusercontent.com/u/852069?v=4" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [kbrandwijk](https://github.com/kbrandwijk)**
_Sunday Jul 23, 2017 at 01:16 GMT_
_Originally opened as https://github.com/graphcool/prisma/issues/324_

----

Currently, there's no way to define permissions for resolver function queries and mutations.

**Comment by marktani** [comment]
<a href="https://github.com/marktani"><img src="https://avatars1.githubusercontent.com/u/1780597?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [marktani](https://github.com/marktani)**
_Monday Jul 24, 2017 at 09:41 GMT_

----

How should that look like in your opinion? The `EVERYONE` and `AUTHENTICATED` permissions would be straight-forward, but how about permission queries?

I'd like to brainstorm about possible PQs and what variables would be needed as a result 🙂 
`$user_id` makes a lot of sense here, what do you think about having access to the scalar input values a well?

**Comment by marktani** [comment]
<a href="https://github.com/kbrandwijk"><img src="https://avatars1.githubusercontent.com/u/852069?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [kbrandwijk](https://github.com/kbrandwijk)**
_Monday Jul 24, 2017 at 09:48 GMT_

----

I already took the first part into consideration when I came up with [this](https://github.com/graphcool/feature-requests/issues/314#issuecomment-316784869). So this fits in there nicely.

Regarding fields, yes I think that it would be the most consistent behavior if input values are available to the PQ.

Also, regarding the PQ themselves. If I have a SE that acts on an existing type (`highlightPost(postId: ID!)`), it makes perfect sense to use `SomePostExist` in a PQ, so I think they can for the most part behave like the other PQ's. Especially with the calling context available (can't find the FR), having `$user_id` in there starts making sense too!

**Comment by marktani** [comment]
<a href="https://github.com/danielrasmuson"><img src="https://avatars2.githubusercontent.com/u/4013767?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [danielrasmuson](https://github.com/danielrasmuson)**
_Wednesday Dec 27, 2017 at 04:55 GMT_

----

What about using the same permission syntax as a check in order to call the function.

I.E.

```
- operation: MyFunctionName
  authenticated: true
  query: ./should-call-my-func.graphql
```

And if that query passes the function will be invoked.

**Comment by marktani** [comment]
<a href="https://github.com/mikhailvs"><img src="https://avatars1.githubusercontent.com/u/506067?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [mikhailvs](https://github.com/mikhailvs)**
_Wednesday Jan 17, 2018 at 03:24 GMT_

----

This feature would be helpful to have.

**Comment by marktani** [comment]
<a href="https://github.com/donedgardo"><img src="https://avatars2.githubusercontent.com/u/2483536?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [donedgardo](https://github.com/donedgardo)**
_Wednesday Jan 17, 2018 at 20:49 GMT_

----

+1
Currently permissions work for crud operations or has this change with v1.0?

**Comment by marktani** [comment]
<a href="https://github.com/LFP6"><img src="https://avatars2.githubusercontent.com/u/18635705?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [LFP6](https://github.com/LFP6)**
_Wednesday Jan 17, 2018 at 21:56 GMT_

----

FWIW this is how permissions are supposed to be handled in Prisma at this point: https://www.prismagraphql.com/docs/tutorials/graphql-server-development/permissions-thohp1zaih Is there still a reason to have a separate mechanism for defining permissions?

**Comment by ahsanwtc** [comment]
yeah would be nice to have restricted access to certain resolvers
